lib/mail/renderers/rfc_2822.ex

```diff
@@ -251,30 +251,83 @@ defmodule Mail.Renderers.RFC2822 do
     |> String.pad_leading(2, "0")
   end
 
-  defp reorganize(%Mail.Message{multipart: true} = message) do
+  defp split_attachment_parts(message) do
+    Enum.reduce(message.parts, [[], [], []], fn part, [texts, mixed, inlines] ->
+      cond do
+        match_content_type?(part, ~r/text\/(plain|html)/) ->
+          [[part | texts], mixed, inlines]
+        Mail.Message.is_inline_attachment?(part) ->
+          [texts, mixed, [part | inlines]]
+        true -> # a mixed part - most likely an attachment
+          [texts, [part | mixed], inlines]
+      end
+    end)
+    |> Enum.map(&Enum.reverse/1) # retain ordering
+  end
+
+  @doc """
+  Will organize message parts to conform to expectations on MIME-part order,
+  specifically in the following format:
+
+      start multipart/mixed
+        start multipart/related
+          start multipart/alternative
+            <text and html parts>
+          end multipart/alternative
+          <inline parts>
+        end multipart/related
+        <attachment parts>
+      end multipart/mixed
+
+  Such that:
+
+  - text and html parts will be grouped in a `multipart/alternative`;
+  - inline attachments will be postpended and grouped with text and html parts
+    in a `multipart/related` (RFC 2387); and
+  - regular attachments will be postpended and grouped with all content in a
+    `multipart/mixed`
+  """
+  def reorganize(%Mail.Message{multipart: true} = message) do
     content_type = Mail.Message.get_content_type(message)
 
-    if Mail.Message.has_attachment?(message) do
-      text_parts =
-        Enum.filter(message.parts, &match_content_type?(&1, ~r/text\/(plain|html)/))
-        |> Enum.sort(&(&1 > &2))
+    [text_parts, mixed, inlines] = split_attachment_parts(message)
+    has_inline = Enum.any?(inlines)
```

Comment:
Might it not be more explicit to use `!Enum.empty?(list)` because the behaviour of `Enum.any?/1` and `Enum.all?/1` are not necessarily intuitive when providing empty lists.